### PR TITLE
fix(core): fix the key/value differ

### LINKED ISSUE #15457: Ng style doesnt remove old style when the model object changes
**Current behaviour** - If you assign an object to ngStyle in a div, ngStyle doesnt flush the style properties from the older object when it changes.

**Expected/desired behavior** - All the old object style should be flushed, and only the new object style properties should be applied.

**Reproduction of the problem**
http://plnkr.co/edit/ccS4gOZgdMqguwIcGRob?p=preview

In the above plunkr, click on first button (margin border background), then on second and then on third. Till this, it works fine.
But now click on second button (border background), margin property should be removed but it isn't.

I had earlier reported the same issue - https://github.com/angular/angular/issues/10315

### PR DESCRIPTION
**Pull request by vicb**
fixes #15457

### CODE REVIEW

**Comment by vicb** [comment]
TODO: optimize the case when the key does not change

File: packages/core/src/change_detection/differs/default_keyvalue_differ.ts [hunk 1]
```diff
@@ -90,67 +95,119 @@ export class DefaultKeyValueDiffer<K, V> implements KeyValueDiffer<K, V>, KeyVal
...
+    let appendAfter: KeyValueChangeRecord_<K, V>;
 
     this._forEach(map, (value: any, key: any) => {
-      let newSeqRecord: any;
-      if (oldSeqRecord && key === oldSeqRecord.key) {
-        newSeqRecord = oldSeqRecord;
-        this._maybeAddToChanges(newSeqRecord, value);
+      if (insertBefore && insertBefore.key === key) {
+        this._maybeAddToChanges(insertBefore, value);
+        [appendAfter, insertBefore] = [insertBefore, insertBefore._next];
```
Comment (mhevery) [inline packages/core/src/change_detection/differs/default_keyvalue_differ.ts]: Don't be clever: http://www.typescriptlang.org/play/#src=%20%5BappendAfter%2C%20insertBefore%5D%20%3D%20%5BinsertBefore%2C%20insertBefore._next%5D%3B It allocates array.

File: packages/core/src/change_detection/differs/default_keyvalue_differ.ts [hunk 2]
```diff
@@ -90,67 +95,119 @@ export class DefaultKeyValueDiffer<K, V> implements KeyValueDiffer<K, V>, KeyVal
...
-          newSeqRecord = records.get(key);
-          this._maybeAddToChanges(newSeqRecord, value);
-        } else {
-          newSeqRecord = new KeyValueChangeRecord_<K, V>(key);
-          records.set(key, newSeqRecord);
-          newSeqRecord.currentValue = value;
-          this._addToAdditions(newSeqRecord);
-        }
+        const record = this._getOrCreateRecordForKey(key, value);
+        [appendAfter, insertBefore] = this._insertBeforeOrAppend(insertBefore, appendAfter, record);
```
Comment (mhevery) [inline packages/core/src/change_detection/differs/default_keyvalue_differ.ts]: This allocates array which creates memory pressure. Can we do it without garbage?

File: packages/core/src/change_detection/differs/default_keyvalue_differ.ts [hunk 3]
```diff
@@ -90,67 +95,119 @@ export class DefaultKeyValueDiffer<K, V> implements KeyValueDiffer<K, V>, KeyVal
...
+      this._removalsHead = insertBefore;
+      this._removalsTail = insertBefore;
+
+      for (let record = insertBefore; record !== null; record = record._nextRemoved) {
+        if (record === this._mapHead) {
+          this._mapHead = null;
         }
+        this._records.delete(record.key);
+        record._nextRemoved = record._next;
+        [record.previousValue, record.currentValue] = [record.currentValue, null];
```
Comment (mhevery) [inline packages/core/src/change_detection/differs/default_keyvalue_differ.ts]: also too clever here.

**Review by mhevery — CHANGES_REQUESTED** [review]

**Review by mhevery — APPROVED** [review]

**Pull request merged**